### PR TITLE
Support code formatting

### PR DESCRIPTION
**Pull request by ZeroX-DG**
This PR aim to support code formatting for the current open file.

### CODE REVIEW

File: src/lspc.rs [hunk 1]
```diff
@@ -131,6 +137,7 @@ pub trait Editor: 'static {
     ) -> Result<(), EditorError>;
     fn show_message(&self, show_message_params: &ShowMessageParams) -> Result<(), EditorError>;
     fn goto(&self, location: &Location) -> Result<(), EditorError>;
+    fn apply_edit(&self, edits: &Vec<TextEdit>) -> Result<(), EditorError>;
```
Comment (unrealhoang) [inline src/lspc.rs]: should be `apply_edits`

**Review by unrealhoang — COMMENTED** [review]
looks good :+1:

File: src/neovim.rs [hunk 2]
```diff
@@ -528,6 +574,53 @@ impl Editor for Neovim {
 
         Ok(())
     }
+
+    fn apply_edits(&self, lines: &Vec<String>, edits: &Vec<TextEdit>) -> Result<(), EditorError> {
+        let mut sorted_edits = edits.clone();
+        let mut editted_content = lines.join("\n");
+        sorted_edits.sort_by(|a, b| match a.range.start.line.cmp(&b.range.start.line) {
```
Comment (unrealhoang) [inline src/neovim.rs]: nit: a trick to do this more clean:
```
sorted_edits.sort_by_key(|i| (i.range.start.line, i.range.start.character));
```

File: src/neovim.rs [hunk 3]
```diff
@@ -528,6 +574,53 @@ impl Editor for Neovim {
...
+            std::cmp::Ordering::Equal => a.range.start.character.cmp(&b.range.start.character),
+            other => other,
+        });
+
+        let mut last_modified_offset = editted_content.len();
+        for i in (0..sorted_edits.len()).rev() {
```
Comment (unrealhoang) [inline src/neovim.rs]: nit
```
for edit in sorted_edits.iter().rev()
```

File: src/neovim.rs (hunk 2 above)
Comment (ZeroX-DG) [reply src/neovim.rs]: That's a cool trick! Thanks!

File: src/neovim.rs (hunk 3 above)
Comment (ZeroX-DG) [reply src/neovim.rs]: Fixed :+1:

File: src/lspc.rs [hunk 4]
```diff
@@ -51,6 +54,12 @@ pub enum Event {
         lang_id: String,
         text_document: TextDocumentIdentifier,
     },
+    FormatDoc {
+        lang_id: String,
+        config: LsConfig,
```
Comment (unrealhoang) [inline src/lspc.rs]: config should be per lang server. i.e. sent to lspc on `start_lang_server`

Comment (unrealhoang) [reply src/lspc.rs]: We should persist it inside struct `LangServerHandler`.

Comment (ZeroX-DG) [reply src/lspc.rs]: A bit weird to persist there but it's certainly better than parsing the settings every time we format the code. Fixed :+1:

File: src/neovim.rs [hunk 5]
```diff
@@ -138,18 +140,57 @@ pub fn from_value(config_value: &Value) -> Option<LsConfig> {
...
+        } else if k == "indentation_with_space" {
+			indentation_with_space = v.as_bool()?;
+		}
     }
     if let (Some(root_markers), Some(command)) = (root_markers, command) {
         Some(LsConfig {
             root_markers,
             command,
+            indentation,
+			indentation_with_space
```
Comment (unrealhoang) [inline src/neovim.rs]: miss formatted here, did you run formatting :joy:

Comment (ZeroX-DG) [reply src/neovim.rs]: :joy: working on formatting feature and forgot to format :rofl:

Comment (ZeroX-DG) [reply src/neovim.rs]: Fixed :+1:

**Pull request merged**